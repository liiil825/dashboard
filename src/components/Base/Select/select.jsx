import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { withTranslation } from 'react-i18next';
import _, { isFunction } from 'lodash';
import { Icon } from 'components/Base';

import Option from './option';

import styles from './index.scss';

@withTranslation()
export default class Select extends React.Component {
  static propTypes = {
    children: PropTypes.any,
    className: PropTypes.string,
    defaultValue: PropTypes.string,
    disabled: PropTypes.bool,
    name: PropTypes.string,
    onChange: PropTypes.func,
    positionFixed: PropTypes.bool,
    value: PropTypes.string
  };

  static defaultProps = {
    className: '',
    disabled: false,
    positionFixed: false,
    name: ''
  };

  state = {
    isOpen: false,
    value: this.props.defaultValue,
    noValue: this.noValue
  };

  componentWillUnmount() {
    document.removeEventListener('click', this.handleOutsideClick.bind(this));
  }

  get value() {
    return this.props.value || this.state.value;
  }

  get noValue() {
    const defaultValue = this.props.value || this.props.defaultValue;
    return typeof defaultValue === 'undefined' || defaultValue === null;
  }

  childNodes = [];

  currentLabel = '';

  handleOutsideClick(e) {
    if (this.wrapper && !this.wrapper.contains(e.target)) {
      this.setState({ isOpen: false });
    }
  }

  handleOptionClick = e => {
    const { onChange, value } = this.props;
    const curVal = e.currentTarget.getAttribute('value');
    const state = {
      isOpen: false,
      value: curVal
    };

    this.setState(state, () => {
      if (isFunction(onChange) && value !== curVal) {
        onChange(curVal);
      }
    });
  };

  handleControlClick = () => {
    const { isOpen } = this.state;
    const { disabled } = this.props;

    if (disabled) {
      return;
    }
    if (isOpen) {
      document.removeEventListener('click', this.handleOutsideClick.bind(this));
    } else {
      document.addEventListener('click', this.handleOutsideClick.bind(this));
    }

    this.setState({ isOpen: !isOpen });
  };

  setChildNodes = () => {
    const { children, t } = this.props;
    const value = this.props.value || this.state.value;

    this.currentLabel = '';
    this.childNodes = React.Children.map(children, child => {
      const checked = String(value) === child.props.value;

      if (checked) {
        this.currentLabel = child.props.children;
      }

      return React.cloneElement(child, {
        ...child.props,
        onClick: this.handleOptionClick,
        isSelected: checked
      });
    });

    if (this.state.noValue) {
      const txt = t('Please select');
      const isSelected = this.noValue;
      const option = React.cloneElement(
        <Option disabled key="please-select">
          {txt}
        </Option>,
        {
          isSelected
        }
      );
      if (isSelected) {
        this.currentLabel = txt;
      }

      this.childNodes.unshift(option);
    }
    if (!this.currentLabel) {
      this.currentLabel = value;
    }

    return this.childNodes;
  };

  renderControl() {
    const { isOpen } = this.state;
    const { disabled } = this.props;

    return (
      <div
        ref={ref => {
          this.control = ref;
        }}
        className={styles.control}
        onClick={this.handleControlClick}
      >
        <div className={styles.controlLabel}>{this.currentLabel}</div>
        <Icon
          name={isOpen && !disabled ? 'caret-up' : 'caret-down'}
          type="dark"
        />
      </div>
    );
  }

  renderOptions() {
    const { isOpen } = this.state;
    const { disabled, positionFixed } = this.props;
    const style = {
      position: positionFixed ? 'fixed' : 'absolute'
    };
    if (positionFixed) {
      const dimensions = this.control
        ? this.control.getBoundingClientRect()
        : null;
      Object.assign(style, {
        width: _.get(dimensions, 'width', 'auto')
      });
    }

    return this.childNodes.length > 0 ? (
      <div
        className={classnames(styles.options, {
          [styles.show]: isOpen && !disabled
        })}
        style={style}
      >
        {this.childNodes}
      </div>
    ) : null;
  }

  render() {
    const { className, disabled, name } = this.props;
    const { value } = this;

    this.setChildNodes();

    return (
      <div
        className={classnames(styles.select, className)}
        disabled={disabled}
        ref={ref => {
          this.wrapper = ref;
        }}
      >
        <input name={name} defaultValue={value} type="hidden" />
        {this.renderControl()}
        {this.renderOptions()}
      </div>
    );
  }
}
