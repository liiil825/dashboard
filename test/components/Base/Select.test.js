import Select from 'components/Base/Select';

const { Option } = Select;

describe('Base/Select', () => {
  let props = {};
  let items = [
    {
      value: 'item1'
    },
    {
      value: 'item2',
      isSelected: true
    },
    {
      value: 'item3'
    }
  ];

  const getWrapper = func => func(
      <Select {...props}>
        {items.map(({ value, ...restProps }) => (
          <Option key={value} {...restProps}>
            {value}
          </Option>
        ))}
      </Select>
  );

  beforeAll(() => {
    props = { ...props };
    items = [...items];
  });

  it('basic render', () => {
    const wrapper = getWrapper(render);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('disabled render', () => {
    props.disabled = true;
    const wrapper = getWrapper(render);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('base mount', () => {
    const spyOnChange = jest.fn();
    props.onChange = spyOnChange;
    const wrapper = getWrapper(mount);
    expect(wrapper.state('isOpen')).toBe(false);
    wrapper.unmount();
  });
});
