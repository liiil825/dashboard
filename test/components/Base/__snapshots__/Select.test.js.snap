// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Select basic render 1`] = `
<div
  class="select"
>
  <input
    name=""
    type="hidden"
  />
  <div
    class="control"
  >
    <div
      class="controlLabel"
    />
    <span
      class="icon icon-caret-down"
      style="width:16px;height:16px"
    >
      <svg
        class="qicon qicon-caret-down qicon-dark"
      >
        <use
          href="#qui-caret-down"
        />
      </svg>
    </span>
  </div>
  <div
    class="options"
  >
    <div
      class="option"
    >
      item1
    </div>
    <div
      class="option"
    >
      item2
    </div>
    <div
      class="option"
    >
      item3
    </div>
  </div>
</div>
`;

exports[`Base/Select disabled render 1`] = `
<div
  class="select"
  disabled=""
>
  <input
    name=""
    type="hidden"
  />
  <div
    class="control"
  >
    <div
      class="controlLabel"
    />
    <span
      class="icon icon-caret-down"
      style="width:16px;height:16px"
    >
      <svg
        class="qicon qicon-caret-down qicon-dark"
      >
        <use
          href="#qui-caret-down"
        />
      </svg>
    </span>
  </div>
  <div
    class="options"
  >
    <div
      class="option"
    >
      item1
    </div>
    <div
      class="option"
    >
      item2
    </div>
    <div
      class="option"
    >
      item3
    </div>
  </div>
</div>
`;
